File: tests/fixtures/classes_structs.go
// Go test fixture for struct and interface parsing.
// Tests various struct and interface declaration styles.
package main

import (
	"fmt"
	"math"
)

// Simple struct
type Point struct {
	X int
	Y int
}

// Struct with tags
type User struct {
	ID       int    `json:"id" db:"user_id"`
	Name     string `json:"name" db:"user_name"`
	Email    string `json:"email" db:"email"`
	IsActive bool   `json:"is_active" db:"active"`
}

// Struct with embedded struct
type Employee struct {
	User
	Department string
	Salary     float64
}

// Interface definition
type Animal interface {
	Speak() string
	Move() string
}

// Interface with multiple methods
type Shape interface {
	Area() float64
	Perimeter() float64
}

// Interface embedding
type ReadWriter interface {
	Reader
	Writer
}

type Reader interface {
	Read(p []byte) (n int, err error)
}

type Writer interface {
	Write(p []byte) (n int, err error)
}

// Struct implementing interface
type Dog struct {
	Name  string
	Breed string
}

func (d Dog) Speak() string {
	return fmt.Sprintf("%s barks", d.Name)
}

func (d Dog) Move() string {
	return fmt.Sprintf("%s runs", d.Name)
}

// Struct with methods
type Rectangle struct {
	Width  float64
	Height float64
}

func (r Rectangle) Area() float64 {
	return r.Width * r.Height
}

func (r Rectangle) Perimeter() float64 {
	return 2 * (r.Width + r.Height)
}

// Struct with pointer receiver methods
type Counter struct {
	value int
}

func (c *Counter) Increment() {
	c.value++
}

func (c *Counter) Decrement() {
	c.value--
}

func (c *Counter) Value() int {
	return c.value
}

// Generic struct (Go 1.18+)
type Container[T any] struct {
	items []T
}

func (c *Container[T]) Add(item T) {
	c.items = append(c.items, item)
}

func (c *Container[T]) Get(index int) T {
	return c.items[index]
}

// Generic interface
type Comparable[T any] interface {
	Compare(other T) int
}

// Empty struct (used for signaling)
type Signal struct{}

// Struct with anonymous fields
type Config struct {
	string
	int
	bool
}

// Type alias
type Celsius float64
type Fahrenheit float64

func (c Celsius) ToFahrenheit() Fahrenheit {
	return Fahrenheit(c*9/5 + 32)
}

// Circle struct for Shape interface
type Circle struct {
	Radius float64
}

func (c Circle) Area() float64 {
	return math.Pi * c.Radius * c.Radius
}

func (c Circle) Perimeter() float64 {
	return 2 * math.Pi * c.Radius
}

func main() {
	p := Point{X: 10, Y: 20}
	fmt.Printf("Point: %+v\n", p)

	dog := Dog{Name: "Buddy", Breed: "Labrador"}
	fmt.Println(dog.Speak())

	rect := Rectangle{Width: 5, Height: 10}
	fmt.Printf("Area: %.2f\n", rect.Area())

	counter := &Counter{}
	counter.Increment()
	counter.Increment()
	fmt.Printf("Counter: %d\n", counter.Value())
}

File: tests/fixtures/classes_structs.rs
/// Rust test fixture for struct and trait parsing.
/// Tests various struct, enum, and trait declaration styles.

// Simple struct
struct Point {
    x: i32,
    y: i32,
}

// Tuple struct
struct Color(u8, u8, u8);

// Unit struct
struct Marker;

// Struct with generics
struct Container<T> {
    item: T,
}

// Struct with lifetime
struct Borrowed<'a> {
    data: &'a str,
}

// Struct with derive macros
#[derive(Debug, Clone, PartialEq)]
struct User {
    id: u64,
    name: String,
    email: String,
}

// Enum (like tagged union)
enum Message {
    Quit,
    Move { x: i32, y: i32 },
    Write(String),
    ChangeColor(u8, u8, u8),
}

// Enum with data
enum Result<T, E> {
    Ok(T),
    Err(E),
}

// Trait definition (like interface)
trait Animal {
    fn speak(&self) -> String;
    fn move_to(&self, x: i32, y: i32);
}

// Trait with default implementation
trait Greet {
    fn name(&self) -> &str;

    fn greet(&self) -> String {
        format!("Hello, {}!", self.name())
    }
}

// Trait with associated type
trait Iterator {
    type Item;
    fn next(&mut self) -> Option<Self::Item>;
}

// Generic trait
trait Comparable<T> {
    fn compare(&self, other: &T) -> i32;
}

// Struct implementing traits
struct Dog {
    name: String,
    breed: String,
}

impl Animal for Dog {
    fn speak(&self) -> String {
        format!("{} barks", self.name)
    }

    fn move_to(&self, x: i32, y: i32) {
        println!("{} moves to ({}, {})", self.name, x, y);
    }
}

impl Greet for Dog {
    fn name(&self) -> &str {
        &self.name
    }
}

// Struct with methods
impl Point {
    // Associated function (constructor)
    fn new(x: i32, y: i32) -> Self {
        Point { x, y }
    }

    // Method
    fn distance(&self, other: &Point) -> f64 {
        let dx = (self.x - other.x) as f64;
        let dy = (self.y - other.y) as f64;
        (dx * dx + dy * dy).sqrt()
    }

    // Mutable method
    fn translate(&mut self, dx: i32, dy: i32) {
        self.x += dx;
        self.y += dy;
    }
}

impl<T> Container<T> {
    fn new(item: T) -> Self {
        Container { item }
    }

    fn get(&self) -> &T {
        &self.item
    }

    fn into_inner(self) -> T {
        self.item
    }
}

// Struct with public fields
pub struct PublicStruct {
    pub field1: i32,
    pub field2: String,
}

// Newtype pattern
struct Meters(f64);
struct Kilometers(f64);

impl Meters {
    fn to_kilometers(&self) -> Kilometers {
        Kilometers(self.0 / 1000.0)
    }
}

fn main() {
    let p1 = Point::new(0, 0);
    let p2 = Point::new(3, 4);
    println!("Distance: {}", p1.distance(&p2));

    let dog = Dog {
        name: String::from("Buddy"),
        breed: String::from("Labrador"),
    };
    println!("{}", dog.speak());
    println!("{}", dog.greet());

    let color = Color(255, 128, 0);
    let container = Container::new(42);
    println!("Container: {}", container.get());

    let user = User {
        id: 1,
        name: String::from("Alice"),
        email: String::from("alice@example.com"),
    };
    println!("User: {:?}", user);
}
